src/mito2/src/compaction/window.rs

```diff
@@ -0,0 +1,380 @@
+            sst_layer: access_layer.clone(),
+            outputs,
+            expired_ssts,
+            compaction_time_window: Some(time_window),
+            request_sender,
+            waiters,
+            file_purger,
+            start_time,
+            cache_manager,
+            storage: current_version.options.storage.clone(),
+            index_options: current_version.options.index_options.clone(),
+            append_mode: current_version.options.append_mode,
+            manifest_ctx,
+            version_control,
+            listener,
+        };
+        Some(Box::new(task))
+    }
+}
+
+fn build_output(
+    windows: BTreeMap<i64, Vec<FileHandle>>,
+    window_size: i64,
+) -> Vec<CompactionOutput> {
+    let mut outputs = Vec::with_capacity(windows.len());
+    for (window_lower_bound, files) in windows {
+        // safety: the upper bound must > lower bound.
+        let output_time_range = Some(
+            TimestampRange::new(
+                Timestamp::new_second(window_lower_bound),
+                Timestamp::new_second(window_lower_bound + window_size),
+            )
+            .unwrap(),
+        );
+
+        let output = CompactionOutput {
+            output_file_id: FileId::random(),
+            output_level: 1,
+            inputs: files,
+            filter_deleted: false,
+            output_time_range,
+        };
+        outputs.push(output);
+    }
+
+    outputs
+}
+
+/// Assigns files to time windows. If file does not contain a time range in metadata, it will be
+/// assigned to a special bucket `i64::MAX` (normally no timestamp can be aligned to this bucket)
+/// so that all files without timestamp can be compacted together.
+fn assign_files_to_time_windows<'a>(
+    bucket_sec: i64,
+    files: impl Iterator<Item = &'a FileHandle>,
+) -> BTreeMap<i64, Vec<FileHandle>> {
+    let mut buckets = BTreeMap::new();
+
+    for file in files {
+        if file.compacting() {
+            continue;
+        }
+        let (start, end) = file.time_range();
+        let bounds = file_time_bucket_span(
+            start.convert_to(TimeUnit::Second).unwrap().value(),
+            end.convert_to(TimeUnit::Second).unwrap().value(),
+            bucket_sec,
+        );
+        for bound in bounds {
+            buckets
+                .entry(bound)
+                .or_insert_with(Vec::new)
+                .push(file.clone());
+        }
+    }
+    buckets
+}
+
+/// Calculates timestamp span between start and end timestamp.
+fn file_time_bucket_span(start_sec: i64, end_sec: i64, bucket_sec: i64) -> Vec<i64> {
+    assert!(start_sec <= end_sec);
+
+    // if timestamp is between `[i64::MIN, i64::MIN.align_by_bucket(bucket)]`, which cannot
+    // be aligned to a valid i64 bound, simply return `i64::MIN` rather than just underflow.
+    let mut start_aligned = start_sec.align_by_bucket(bucket_sec).unwrap_or(i64::MIN);
+    let end_aligned = end_sec.align_by_bucket(bucket_sec).unwrap_or(i64::MIN);
```

Comment:
Should we handle overflow by using falling back to `i64::MIN` and `i64::MAX`? They are not aligned by the `bucket_sec` so we might get an incorrect window if we add `bucket_sec` to `start_aligned`.

Maybe we could ignore this file?

Reply:
> Should we handle overflow by using falling back to `i64::MIN` and `i64::MAX`? They are not aligned by the `bucket_sec` so we might get an incorrect window if we add `bucket_sec` to `start_aligned`.
> 
> Maybe we could ignore this file?

`align_by_bucket` will only underflow for those timestamps near `i64::MIN`, so we can simply assign them to the special `i64::MIN` window.